### PR TITLE
 DefaultGenerator: ignore only form param schemas

### PR DESCRIPTION
**Pull request by jmini**
Fix for #50.

This looks for all used schema to compute the unused one.

### CODE REVIEW

**Comment by wing328** [comment]
I tested this one with
```
git checkout -b jmini-issue50 master
git pull https://github.com/jmini/openapi-generator.git issue50
```
and ran `./bin/ruby-petstore.sh` but got the following:
```diff
--- a/samples/client/petstore/ruby/README.md
+++ b/samples/client/petstore/ruby/README.md
@@ -109,38 +109,11 @@ Class | Method | HTTP request | Description
 
 ## Documentation for Models
 
- - [Petstore::AdditionalPropertiesClass](docs/AdditionalPropertiesClass.md)
- - [Petstore::Animal](docs/Animal.md)
- - [Petstore::AnimalFarm](docs/AnimalFarm.md)
  - [Petstore::ApiResponse](docs/ApiResponse.md)
- - [Petstore::ArrayOfArrayOfNumberOnly](docs/ArrayOfArrayOfNumberOnly.md)
- - [Petstore::ArrayOfNumberOnly](docs/ArrayOfNumberOnly.md)
- - [Petstore::ArrayTest](docs/ArrayTest.md)
- - [Petstore::Capitalization](docs/Capitalization.md)
- - [Petstore::Cat](docs/Cat.md)
- - [Petstore::Category](docs/Category.md)
- - [Petstore::ClassModel](docs/ClassModel.md)
  - [Petstore::Client](docs/Client.md)
- - [Petstore::Dog](docs/Dog.md)
- - [Petstore::EnumArrays](docs/EnumArrays.md)
- - [Petstore::EnumClass](docs/EnumClass.md)
- - [Petstore::EnumTest](docs/EnumTest.md)
- - [Petstore::FormatTest](docs/FormatTest.md)
- - [Petstore::HasOnlyReadOnly](docs/HasOnlyReadOnly.md)
- - [Petstore::List](docs/List.md)
- - [Petstore::MapTest](docs/MapTest.md)
- - [Petstore::MixedPropertiesAndAdditionalPropertiesClass](docs/MixedPropertiesAndAdditionalPropertiesClass.md)
- - [Petstore::Model200Response](docs/Model200Response.md)
- - [Petstore::ModelReturn](docs/ModelReturn.md)
- - [Petstore::Name](docs/Name.md)
- - [Petstore::NumberOnly](docs/NumberOnly.md)
  - [Petstore::Order](docs/Order.md)
  - [Petstore::OuterComposite](docs/OuterComposite.md)
- - [Petstore::OuterEnum](docs/OuterEnum.md)
  - [Petstore::Pet](docs/Pet.md)
- - [Petstore::ReadOnlyFirst](docs/ReadOnlyFirst.md)
- - [Petstore::SpecialModelName](docs/SpecialModelName.md)
- - [Petstore::Tag](docs/Tag.md)
  - [Petstore::User](docs/User.md)
```

A lot of models are no longer auto-generated.

**Comment by jmini** [comment]
Ok...
@wing328 thank you a lot for this test.

I did not check in detail, but I guess that those model definitions are not used in the [petstore-with-fake-endpoints-models-for-testing.yaml](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/2_0/petstore-with-fake-endpoints-models-for-testing.yaml)

---

I still do not get the precise requirements are for this `getUnusedSchemas()` method and why It was added in the first place.

The point, if you work with the OAS2 proposed in #50, some Model classes are missing (code is not compiling) so this is a regression compared to Swagger-Codegen 2.x. @etherealjoy has also reported a regression with "aspnetcore" generator.

We need to do something.

**Comment by jimschubert** [comment]
Something to consider as well, some folks (like me) generate models from models-only specs. Will this break that use case? If so, I think we'll need to make it conditional.

**Comment by wing328** [comment]
> Something to consider as well, some folks (like me) generate models from models-only specs. Will this break that use case? If so, I think we'll need to make it conditional.

Yes, I'm aware of such use case.

**Comment by jmini** [comment]
@wing328: let summarize it that way:

The current implementation does:
In `DefaultGenerator.generate()` in `unusedSchemas` you call `ModelUtils.getUnusedSchemas(openAPI)` to get the list of schemas used in a body-request with MimeType: `application/x-www-form-urlencoded` or `multipart/form-data`.

I think what the correct implementation should do is:
get the list of schemas used in a body-request with MimeType: `application/x-www-form-urlencoded` or `multipart/form-data` **that are not used somewhere else** (for example in a json mime type as it is the case in my issue #50)

If you confirm this, I can rework the PR base on this requirement.

**Comment by jmini** [comment]
Input @wing328 :
> in openapi v2 spec, there's form parameter
but in openapi v3 spec, it becomes part of the requestbody
and in our code, we've used the flatten feature in openapi parser v3
which means those form parameters defined in the requestbody will be generated as models
these models are not necessary
developers upgrading their spec to v3 will find a lot of models generated but not defined by them in the component section
that's why i check the MIME type to see if it's for form parameters
and removed those if found

---

I think that the `ModelUtils.getSchemasUsedOnlyInFormParam(OpenAPI)` now does what it should (according to what was discussed here).

**Comment by jmini** [comment]
@wing328 I have tested it with `./bin/ruby-petstore.sh` and for me, it now works as expected. Can you review?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java [hunk 1]
```diff
@@ -85,49 +88,128 @@ public static CodegenModel getModelByName(final String name, final Map<String, O
...
+    /**
+                                    if (!"application/x-www-form-urlencoded".equalsIgnoreCase(mimeType) && 
+                                        !"multipart/form-data".equalsIgnoreCase(mimeType)) {
+                                    // remove the schema that's automatically created by the parser
+                                    MediaType mediaType = requestBody.getContent().get(mimeType);
+                                    if (mediaType.getSchema() != null && mediaType.getSchema().get$ref() != null) {
+                                        unusedSchemas.remove(getSimpleRef(mediaType.getSchema().get$ref()));
+                                    }
+                                }
+     */
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java]: @jmini what about removing this commented code block?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java [hunk 2]
```diff
@@ -85,49 +88,128 @@ public static CodegenModel getModelByName(final String name, final Map<String, O
...
+
+    private static void visitOpenAPI(OpenAPI openAPI, OpenAPISchemaVisitor visitor) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java]: I would suggest adding a 1-liner explaining what this function does.

**Comment by jmini** [comment]
@wing328 nice catch, I have removed the dead code and add a javadoc on the private method.

Thank you a lot for this review.

**Pull request merged**